### PR TITLE
[sb2] Replace deprecated egrep, fgrep and cleanup

### PR DESCRIPTION
**Pull request by Thaodan**

### CODE REVIEW

File: utils/sb2 [hunk 1]
```diff
@@ -168,8 +168,8 @@ function locate_target_nsswitch_conf()
 		-L $SBOX_LIBSB2 -- $SBOX_DIR/bin/sb2-show \
 		which /var/run/nscd/socket \
 		>$SBOX_SESSION_DIR/path_to_nscd_socket.conf
-	NSSWITCH_CONF_PATH=`cat $SBOX_SESSION_DIR/path_to_nsswitch.conf`
-	NSCD_SOCKET_PATH=`cat $SBOX_SESSION_DIR/path_to_nscd_socket.conf`
+    read -r NSSWITCH_CONF_PATH < $SBOX_SESSION_DIR/path_to_nsswitch.conf
+    read -r NSCD_SOCKET_PATH < $SBOX_SESSION_DIR/path_to_nscd_socket.conf
```
Comment (mlehtima) [inline utils/sb2]: Indentation mismatch, also some other places have similar, recheck the indentations in your changes.

File: utils/sb2 [hunk 2]
```diff
@@ -655,12 +654,12 @@ function write_ld_library_path_replacement_to_exec_config()
...
 	# Include directories listed in ld.so.conf
 	if [ -f $rootdir/etc/ld.so.conf ]; then
-		lloc2=`egrep '^/' $rootdir/etc/ld.so.conf`
+		lloc2=$(grep -E '^/' $rootdir/etc/ld.so.conf)
 		liblocations="$liblocations $lloc2"
 	fi
 	# Include directories listed in ld.so.conf.d/*
 	if [ -d $rootdir/etc/ld.so.conf.d ]; then
-		lloc2=`cat $rootdir/etc/ld.so.conf.d/* 2>/dev/null | egrep '^/'`
+		lloc2=$(grep --extended-regexp --only-matching '^/' $rootdirq/etc/ld.so.conf.d/* )
```
Comment (mlehtima) [inline utils/sb2]: Typo rootdirq? Should that be rootdir.

File: utils/sb2-upgrade-config [hunk 3]
```diff
@@ -21,56 +21,54 @@ SBOX_DIR=$(readlink -f $SBOX_SHARE_DIR/../..)
 SBOX_TARGET=$1
 SBOX_CONFIG_DIR=~/.scratchbox2/$SBOX_TARGET/sb2.config.d 
 
-function exit_error()
+exit_error()
 {
-	echo "$my_path: Error: $@"
+	echo "$my_path: Error: " "$@"
```
Comment (mlehtima) [inline utils/sb2-upgrade-config]: Why this change?

**Review by mlehtima — CHANGES_REQUESTED** [review]
Also one of the commit messages mentions both egrep and fgrep but in that commit only egrep changes have been made.

File: utils/sb2-upgrade-config (hunk 3 above)
Comment (Thaodan) [reply utils/sb2-upgrade-config]: Because the way it expands. `$@` is an array which will expand into `$1`, `$2` etc,
which means each counts as it's own argument if this behavior is intended the string
and the following `$@` should be quoted separately or if they should be combined
into one string `$*` fits better.

I changed it to the latter now.

File: utils/sb2 [hunk 4]
```diff
@@ -669,11 +668,11 @@ function write_ld_library_path_replacement_to_exec_config()
...
-		dirs_in_cache=`$rootdir/sbin/ldconfig -p \
-			-C $rootdir/etc/ld.so.cache | 
-			fgrep '=>' | 
-			sed -e 's/^.*=> //' -e 's:/[^/]*$::' | 
-			sort | uniq`
+		dirs_in_cache=$($rootdir/sbin/ldconfig --print-cache \
+											   -r "$rootdir" \
+											   -C "$rootdir/etc/ld.so.cache" |
+							sed -En '/=>/ {s/.*=> (.*)/\1/;s|(.*)/.*$|\1|p}' |
+			                sort | uniq)
```
Comment (mlehtima) [inline utils/sb2]: Indentation here is a mess, please fix.

File: utils/sb2 [hunk 5]
```diff
@@ -728,10 +727,11 @@ function check_qemu_features()
 	fi
 
 	# run qemu -h
+	read qemu_cmdline < $SBOX_SESSION_DIR/qemu_cmdline
 	__SB2_BINARYNAME="sb2:CheckingQemuFeatures" \
 	sb2-monitor \
 		-L $SBOX_LIBSB2 -- \
-		`cat $SBOX_SESSION_DIR/qemu_cmdline` -h \
+	    $qemu_cmdline -h \
```
Comment (mlehtima) [inline utils/sb2]: Indentation mismatch.

File: utils/sb2-upgrade-config [hunk 6]
```diff
@@ -21,56 +21,54 @@ SBOX_DIR=$(readlink -f $SBOX_SHARE_DIR/../..)
...
 
 if [ -z "$SBOX_TARGET" ]; then
 	exit_error "this script is intended for sb2's internal use only!"
 fi 
 
-function log_config_action()
+log_config_action()
 {
-	tstamp=`/bin/date '+%Y-%m-%d %H:%M:%S'`
+	tstamp=$(/bin/date '+%Y-%m-%d %H:%M:%S')
```
Comment (mlehtima) [inline utils/sb2-upgrade-config]: Do we need /bin prefix in this?

File: utils/sb2 [hunk 7]
```diff
@@ -655,12 +654,12 @@ function write_ld_library_path_replacement_to_exec_config()
...
 	# Include directories listed in ld.so.conf
 	if [ -f $rootdir/etc/ld.so.conf ]; then
-		lloc2=`egrep '^/' $rootdir/etc/ld.so.conf`
+		lloc2=$(grep -E '^/' $rootdir/etc/ld.so.conf)
 		liblocations="$liblocations $lloc2"
 	fi
 	# Include directories listed in ld.so.conf.d/*
 	if [ -d $rootdir/etc/ld.so.conf.d ]; then
-		lloc2=`cat $rootdir/etc/ld.so.conf.d/* 2>/dev/null | egrep '^/'`
+		lloc2=$(grep --extended-regexp --only-matching '^/' $rootdir/etc/ld.so.conf.d/* )
```
Comment (mlehtima) [inline utils/sb2]: Are we sure this path always exists? The previous code didn't make that assumption.

**Review by mlehtima — CHANGES_REQUESTED** [review]

File: utils/sb2 [hunk 8]
```diff
@@ -669,11 +675,11 @@ function write_ld_library_path_replacement_to_exec_config()
 		# print contents of the cache, take destination
 		# names (the part after "=>"), drop file names (leave
 		# directory names, and remove duplicates:
-		dirs_in_cache=`$rootdir/sbin/ldconfig -p \
-			-C $rootdir/etc/ld.so.cache | 
-			fgrep '=>' | 
-			sed -e 's/^.*=> //' -e 's:/[^/]*$::' | 
-			sort | uniq`
+		dirs_in_cache=$($rootdir/sbin/ldconfig --print-cache \
```
Comment (mlehtima) [inline utils/sb2:678]: Indentation issues in the following lines.

File: utils/sb2 [hunk 9]
```diff
@@ -1352,29 +1360,29 @@ function add_map_mode()
...
 ##	if [ -z "$SBOX_MAPMODE" ]; then
 ##		# SBOX_MAPMODE was not set. Use basename of the rule file
 ##		# as default mode name:
-##		SBOX_MAPMODE=`basename $1 .lua`
+##		SBOX_MAPMODE=$(basename $1 .lua)
 ##	fi
 ##
 ##	SB2_EXTERNAL_RULEFILES="$SB2_EXTERNAL_RULEFILES $1"
-##	NUM_MAPMODES=`expr $NUM_MAPMODES + 1`
+##     NUM_MAPMODES=$(($NUM_MAPMODES + 1))
```
Comment (mlehtima) [inline utils/sb2]: Unnecessary change from tab to spaces causing indentation mismatch.

**Review by mlehtima — CHANGES_REQUESTED** [review]

**Review by mlehtima — APPROVED** [review]
